fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ScalarFunctionValue.java

```diff
@@ -0,0 +1,332 @@
+        return operation.eval(children.stream().map(c -> c.eval(store, context)).collect(Collectors.toList()));
+    }
+
+    @Nonnull
+    @Override
+    public String explain(@Nonnull final Formatter formatter) {
+        return operation.name().toLowerCase(Locale.getDefault()) + "(" + children.stream().map(c -> c.explain(formatter)).collect(Collectors.joining(",")) + ")";
+    }
+
+    @Nonnull
+    @Override
+    public Type getResultType() {
+        return Type.primitiveType(operation.getResultType());
+    }
+
+    @Nonnull
+    @Override
+    public Iterable<? extends Value> getChildren() {
+        return ImmutableList.copyOf(children);
+    }
+
+    @Nonnull
+    @Override
+    public ScalarFunctionValue withChildren(final Iterable<? extends Value> newChildren) {
+        Verify.verify(Iterables.size(newChildren) == 2);
+        return new ScalarFunctionValue(this.operation, ImmutableList.copyOf(newChildren));
+    }
+
+    @Override
+    public int hashCodeWithoutChildren() {
+        return PlanHashable.objectsPlanHash(PlanHashKind.FOR_CONTINUATION, BASE_HASH, operation);
+    }
+    
+    @Override
+    public int planHash(@Nonnull final PlanHashKind hashKind) {
+        return PlanHashable.objectsPlanHash(hashKind, BASE_HASH, operation, children);
+    }
+
+    @Override
+    public String toString() {
+        return operation.name().toLowerCase(Locale.getDefault()) + "(" + children.stream().map(Object::toString).collect(Collectors.joining(",")) + ")";
+    }
+
+    @Override
+    public int hashCode() {
+        return semanticHashCode();
+    }
+
+    @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
+    @SpotBugsSuppressWarnings("EQ_UNUSUAL")
+    @Override
+    public boolean equals(final Object other) {
+        return semanticEquals(other, AliasMap.identitiesFor(getCorrelatedTo()));
+    }
+
+    @Nonnull
+    private static Map<Pair<ScalarFunction, TypeCode>, PhysicalOperator> getOperatorMap() {
+        return operatorMapSupplier.get();
+    }
+
+    @Nonnull
+    @SuppressWarnings("OptionalGetWithoutIsPresent")
+    private static Value encapsulate(@Nonnull BuiltInFunction<Value> builtInFunction,
+                                     @Nonnull final List<? extends Typed> arguments) {
+        Verify.verify(arguments.size() >= 2);
+        Type resultType = arguments.get(0).getResultType();
+        for (final var arg : arguments) {
+            Type argType = arg.getResultType();
+            SemanticException.check(argType.isPrimitive(), SemanticException.ErrorCode.ARGUMENT_TO_ARITHMETIC_OPERATOR_IS_OF_COMPLEX_TYPE);
+            if (resultType.isUnresolved()) {
+                resultType = argType;
+            } else if (!argType.isUnresolved()) {
+                resultType = Type.maximumType(resultType, argType);
+            }
+            SemanticException.check(resultType != null, SemanticException.ErrorCode.INCOMPATIBLE_TYPE);
```

Comment:
This can only trigger if the `else if` one line up was taken. Maybe move this check into that branch.